.github/workflows/clients.yml

```diff
@@ -0,0 +1,109 @@
+name: Clients
+
+concurrency:
+  group: clients-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+on:
+  workflow_call:
+
+jobs:
+  client-dotnet:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest, [self-hosted, ARM64, macos-12.6], [self-hosted, ARM64, macos-13.2]]
+        dotnet-version: ['3.1.x', '7.0.x']
+
+    runs-on: ${{ matrix.os }}
+    permissions:
+      # Our M1 runners are self hosted, and not ephemeral like standard GitHub action runners.
+      # Limit the token permissions to be read - so even if they were compromised, no damage
+      # could be done.
+      contents: read
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-dotnet@v3
+        with:
+          dotnet-version: ${{ matrix.dotnet-version }}
+
+      - run: ./scripts/install_zig.${{ matrix.os == 'windows-latest' && 'bat' || 'sh' }}
+      - run: ./zig/zig build client_ci -- --language=dotnet
+      - run: ./zig/zig build client_docs -- --language=dotnet
+      - if: matrix.os == 'ubuntu-latest'
+        run: ./.github/ci/fail_on_diff.sh
```

Comment:
I want to fold docs into `client_ci`, but that's for a follow up, as I want to do a somewhat large refactor there (basically, the way the code works today is that pieces of reademe are concatenated to form an executable example. I'd rather we have a full example in the repo, and generate the readme by extracting parts of that.